File: frontend/src/api/research.ts
import { getJson } from "./client";

export type ResearchCandidate = {
  name: string;
  code: string;
  industry: string;
  tendency: string;
  confidence: string;
  risk: string;
  as_of_date: string | null;
  core_advantage: string;
  primary_risk: string;
  score: number | null;
  coverage: number | null;
};

export type ResearchHome = {
  status: string;
  reason: string | null;
  freshness: {
    latest_trade_date: string;
    research_snapshot_date: string;
    security_count: number;
    state: string;
    reason: string;
  };
  candidates: ResearchCandidate[];
};

export type CandidatePage = {
  status: string;
  reason: string | null;
  rows: ResearchCandidate[];
  total: number;
  current_page: number;
  page_count: number;
  page_size: number;
};

export type CandidateQuery = Record<string, string | number | undefined>;

export function getResearchHome(): Promise<ResearchHome> {
  return getJson<ResearchHome>("/api/v1/research/home");
}

export function getCandidates(query: CandidateQuery = {}): Promise<CandidatePage> {
  const params = new URLSearchParams();
  for (const [key, value] of Object.entries(query)) {
    if (value !== undefined && value !== "") params.set(key, String(value));
  }
  const suffix = params.size ? `?${params}` : "";
  return getJson<CandidatePage>(`/api/v1/research/candidates${suffix}`);
}

File: backend/app/schemas/research.py
"""Public research API schemas."""

from __future__ import annotations

from pydantic import BaseModel, Field


class ResearchCandidate(BaseModel):
    name: str
    code: str
    industry: str
    tendency: str
    confidence: str
    risk: str
    as_of_date: str | None
    core_advantage: str
    primary_risk: str
    score: float | None
    coverage: float | None


class ResearchFreshness(BaseModel):
    latest_trade_date: str
    research_snapshot_date: str
    security_count: int
    state: str
    reason: str


class ResearchHomeResponse(BaseModel):
    status: str
    reason: str | None
    freshness: ResearchFreshness
    candidates: list[ResearchCandidate]


class ResearchCandidatePageResponse(BaseModel):
    status: str
    reason: str | None
    rows: list[ResearchCandidate]
    total: int = Field(ge=0)
    current_page: int = Field(ge=1)
    page_count: int = Field(ge=1)
    page_size: int = Field(ge=1, le=100)
